compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/IrTypeMapper.kt

```diff
@@ -65,10 +65,10 @@ class IrTypeMapper(private val context: JvmBackendContext) : KotlinTypeMapperBas
                     append(shortName)
                 }
             is IrClass ->
-                return computeClassInternalName(parent).append("$").append(shortName)
+                return StringBuilder(classInternalName(parent)).append("$").append(shortName)
```

Comment:
I suppose the problem started happening after f2b0d057b9ec17686deb5d25990624b378ef3e3a where I've tried to optimize this code to minimize string concatenations and map lookups for deeply nested classes. Is there any way we could keep the optimization and also fix the class names, maybe by storing the class names in the `context.getLocalClassType` map? It'd be great to have names for all local classes in that map, so that only one lookup would be necessary in the type mapper.

Reply:
This type is actually in the localClassType map. The problem is that we do not look it up when we do a recursive call here. Let me try to restructure this so that we do the localClassType map lookup on recursive calls but thread around a StringBuilder so we only build the strings once. I'll benchmark and see how that performs compared to what is there now for some deep hierarchies.

Reply:
Thanks for catching that @udalov. I verified that for deeply nested classes my initial attempt added a 10% overhead on the full compilation time. That is not good and is indeed caused by the string building changing.

I updated to always lookup types in the local class type mapping first in computeClassInternalName. That fixes the issue and has only on the order of ~1 percent compile time penalty.